### PR TITLE
fix(saferefluxstore): do not return from while loop

### PR DESCRIPTION
**Pull request by JonasBa**
This is a bug that prevents other listeners from being cleaned as return will exit the while loop

### CODE REVIEW

**Review by Zylphrex — APPROVED** [review]

**Review by k-fish — APPROVED** [review]
Nice catch.

File: tests/js/spec/views/organizationGroupDetails/groupEventDetailsContainer.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  it('fetches environments', async function () {
+    const environmentsCall = MockApiClient.addMockResponse({
+      url: `/organizations/${organization.slug}/environments/`,
+      body: TestStubs.Environments(),
+    });
+
+    render(<GroupEventDetailsContainer {...makeProps({organization})} />);
+    expect(await screen.findByTestId('loading-indicator')).toBeInTheDocument();
+
+    expect(environmentsCall).toHaveBeenCalledTimes(1);
```
Comment (Zylphrex) [inline tests/js/spec/views/organizationGroupDetails/groupEventDetailsContainer.spec.tsx:51]: Can we add back the assert that it's finished loading as well?

**Review by Zylphrex — APPROVED** [review]

File: tests/js/spec/views/organizationGroupDetails/groupEventDetailsContainer.spec.tsx (hunk 1 above)
Comment (JonasBa) [reply tests/js/spec/views/organizationGroupDetails/groupEventDetailsContainer.spec.tsx:51]: I thought about even just removing the L49 here and just assert that fetch is called

File: static/app/views/organizationGroupDetails/groupEventDetails/index.tsx [hunk 2]
```diff
@@ -31,53 +29,48 @@ type Props = RouteComponentProps<
   organization: Organization;
   project: Project;
   selection: PageFilters;
-};
+}
 
-type State = typeof OrganizationEnvironmentsStore['state'];
+export function GroupEventDetailsContainer(props: GroupEventDetailsProps) {
```
Comment (evanpurkhiser) [inline static/app/views/organizationGroupDetails/groupEventDetails/index.tsx:34]: Any chance you could pull out this FC refactor into it's own PR?

File: static/app/stores/organizationEnvironmentsStore.tsx [hunk 3]
```diff
@@ -89,6 +93,6 @@ const storeConfig: StoreDefinition &
 
 const OrganizationEnvironmentsStore = createStore(
   makeSafeRefluxStore(storeConfig)
-) as Store & OrganizationEnvironmentsStoreInterface;
+) as SafeRefluxStore & OrganizationEnvironmentsStoreInterface;
```
Comment (evanpurkhiser) [inline static/app/stores/organizationEnvironmentsStore.tsx:96]: I noticed we were inconsistent with some of our store types, sometimes we type it as Store other times as SafeStore. Why is that?

Comment (JonasBa) [reply static/app/stores/organizationEnvironmentsStore.tsx:96]: Not all stores are created with the safeRefluxStore, only the ones with the listenTo in their init methods

Comment (evanpurkhiser) [reply static/app/stores/organizationEnvironmentsStore.tsx:96]: Got it!

File: static/app/views/organizationGroupDetails/groupEventDetails/index.tsx (hunk 2 above)
Comment (evanpurkhiser) [reply static/app/views/organizationGroupDetails/groupEventDetails/index.tsx]: To explain a bit why I suggested this --

It’s basically optimizing for writing code vs optimizing to reviewers of the changes. A nice thing about optimizing for review is that if you give people concise small things to look at (that perhaps can add up to a combination of bigger changes, but all separate PRs) you’ll tend to be able to get things merged in faster, vs if you have 2 semi related, but not obviously easy to review together changes, the reviewer has to do extra work to understand the two changes together.

So by pulling the FC conversion / test updates into a single PR we don't also need to think about the fix to the safe reflux store and can approve that quickly. Then after when looking at this safe reflux store fix we don't need to think about the FC/test changes.

Comment (JonasBa) [reply static/app/views/organizationGroupDetails/groupEventDetails/index.tsx]: I dont mind that, as long as we enforce it across the rest of the org. The problem here was that fixing the reflux store breaks the tests, so had to have been updated too. I moved the fc to separate PR tho

Comment (evanpurkhiser) [reply static/app/views/organizationGroupDetails/groupEventDetails/index.tsx]: Separate PR for those looking for it https://github.com/getsentry/sentry/pull/33141

**Comment by JonasBa** [comment]
Checked with @narsaynorath, it seems that this change now at least causes one of the flaky tests to consistently fail in CI, I cannot reproduce locally, but we are deciding to skip that test for now.

File: static/app/views/organizationGroupDetails/groupEventDetails/index.tsx (hunk 2 above)
Comment (k-fish) [reply static/app/views/organizationGroupDetails/groupEventDetails/index.tsx]: @JonasBa it's not so much about enforcement, imo maybe this PR is on the borderline of it's small enough to not matter, but in general I also _try_ to adhere to the the rule of "_1 logical unit of work_" per PR / branch, since it does make reviewing way easier and faster. I know from a reviewing standpoint it makes me (at least) more inclined to review 😆

Comment (JonasBa) [reply static/app/views/organizationGroupDetails/groupEventDetails/index.tsx]: I'm 200% on that train too :D I think we need to do this more and encourage folks to do it. I've personally avoided this process because of the state of our CI, merging tiny PR's was just a nightmare of just endlessly rerunning CI until it passed. But that has improved in this past week so I'll be sure we do that and encourage others to do the same :)
